File: packages/database/src/migrate.ts
import { readFile, readdir } from "node:fs/promises";
import { basename, dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { Pool } from "pg";

const migrationsDirectory = join(
  dirname(fileURLToPath(import.meta.url)),
  "../migrations",
);

export async function migrate(pool: Pool): Promise<void> {
  const migrations = (await readdir(migrationsDirectory))
    .filter((name) => name.endsWith(".sql"))
    .sort();

  const client = await pool.connect();
  try {
    await client.query("BEGIN");
    await client.query(
      "SELECT pg_advisory_xact_lock(hashtext('analysis-tool-schema-migrations'))",
    );
    await client.query(`
      CREATE TABLE IF NOT EXISTS schema_migrations (
        name text PRIMARY KEY,
        applied_at timestamptz NOT NULL DEFAULT now()
      )
    `);

    for (const name of migrations) {
      const applied = await client.query(
        "SELECT 1 FROM schema_migrations WHERE name = $1",
        [name],
      );
      if (applied.rowCount === 0) {
        await client.query(await readFile(join(migrationsDirectory, name), "utf8"));
        await client.query("INSERT INTO schema_migrations (name) VALUES ($1)", [
          name,
        ]);
      }
    }

    await client.query("COMMIT");
  } catch (error) {
    await client.query("ROLLBACK");
    throw error;
  } finally {
    client.release();
  }
}

async function runCli(): Promise<void> {
  const pool = new Pool({ connectionString: process.env.DATABASE_URL });
  try {
    await migrate(pool);
  } finally {
    await pool.end();
  }
}

if (
  process.argv[1] === fileURLToPath(import.meta.url) &&
  basename(process.argv[1]).startsWith("migrate.")
) {
  await runCli();
}

File: apps/ai-worker/src/ai_worker/migrate.py
import os
from pathlib import Path

import psycopg

# apps/ai-worker/src/ai_worker/migrate.py -> parents[4] is the repo root, both
# locally and inside the Docker image (the Dockerfile mirrors this exact
# relative layout so the default needs no override there).
DEFAULT_MIGRATIONS_DIR = (
    Path(__file__).resolve().parents[4] / "packages" / "database" / "migrations"
)


def _migrations_dir() -> Path:
    override = os.environ.get("DATABASE_MIGRATIONS_DIR")
    return Path(override) if override else DEFAULT_MIGRATIONS_DIR


async def migrate(conn: psycopg.AsyncConnection) -> None:
    """Python port of packages/database/src/migrate.ts's algorithm.

    Uses the same advisory-lock name and `schema_migrations` bookkeeping so
    this can safely run concurrently with the TypeScript worker's own
    migrate() call at container startup — whichever wins applies pending
    migrations, the other blocks briefly on the lock and then no-ops.
    """
    migrations_dir = _migrations_dir()
    names = sorted(
        entry.name
        for entry in migrations_dir.iterdir()
        if entry.is_file() and entry.name.endswith(".sql")
    )

    async with conn.transaction():
        async with conn.cursor() as cur:
            await cur.execute(
                "SELECT pg_advisory_xact_lock(hashtext('analysis-tool-schema-migrations'))"
            )
            await cur.execute(
                """
                CREATE TABLE IF NOT EXISTS schema_migrations (
                  name text PRIMARY KEY,
                  applied_at timestamptz NOT NULL DEFAULT now()
                )
                """
            )
            for name in names:
                await cur.execute(
                    "SELECT 1 FROM schema_migrations WHERE name = %s", (name,)
                )
                if await cur.fetchone() is None:
                    sql = (migrations_dir / name).read_text(encoding="utf-8")
                    await cur.execute(sql)
                    await cur.execute(
                        "INSERT INTO schema_migrations (name) VALUES (%s)", (name,)
                    )
